### PR TITLE
Fix 130: check next_valid_states when changing SubmissionStates

### LINKED ISSUE #130: Use `next_valid_states` information on Submission changes
On the `SubmissionTypes` class, there is a "next valid states" lookup which should be used in any method like `accept()`.

### PR DESCRIPTION
**Pull request by luto**
fixes #130 :)

### CODE REVIEW

**Comment by luto** [comment]
@rixx rebased. take a look please, if you want :)

File: src/pretalx/submission/models/submission.py [hunk 1]
```diff
@@ -180,20 +181,36 @@ def get_duration(self):
...
+            self.save(update_fields=['state'])
+        else:
+            source_states = (src for src, dsts in SubmissionStates.valid_next_states.items() if new_state in dsts)
+
+            # build an error message mentioning all states, which are valid source states for the desired new state.
+            trans_or = pgettext('in talk state errors: "... must be accepted OR foo OR bar ..."', ' or ')
+            state_names = dict(SubmissionStates.get_choices())
+            source_states = trans_or.join(str(state_names[state]) for state in source_states)
+            raise SubmissionError(
+                _('Submission must be {src_states} not {state} to be confirmed.').format(
```
Comment (rixx) [inline src/pretalx/submission/models/submission.py]: "… to be set to {new_state}", I think.

File: src/pretalx/submission/models/submission.py [hunk 2]
```diff
@@ -180,20 +181,36 @@ def get_duration(self):
...
+        valid_next_states = SubmissionStates.valid_next_states[self.state]
+
+        if new_state in valid_next_states or force:
+            self.state = new_state
+            self.save(update_fields=['state'])
+        else:
+            source_states = (src for src, dsts in SubmissionStates.valid_next_states.items() if new_state in dsts)
+
+            # build an error message mentioning all states, which are valid source states for the desired new state.
+            trans_or = pgettext('in talk state errors: "... must be accepted OR foo OR bar ..."', ' or ')
```
Comment (rixx) [inline src/pretalx/submission/models/submission.py]: Nice! (Although 'in talk state errors:' is a bit hard to parse.)

File: src/pretalx/submission/models/submission.py [hunk 3]
```diff
@@ -180,20 +181,36 @@ def get_duration(self):
...
-    def confirm(self, person=None, force=False, orga=False):
-        if self.state not in [SubmissionStates.ACCEPTED] and not force:
-            raise SubmissionError(_('Submission must be accepted, not {state} to be confirmed.').format(state=self.state))
+    def _set_state(self, new_state, force=False):
+        """
+        check if the new state is valid for this Submission (based on SubmissionStates.valid_next_states).
+        if yes, set it and save the object. if no, raise a SubmissionError with a helpful message.
+        """
+
+        valid_next_states = SubmissionStates.valid_next_states[self.state]
```
Comment (rixx) [inline src/pretalx/submission/models/submission.py]: Just to be really safe, I'd prefer a `.get(self.state, [])` here – if the db is corrupted, somebody sets a wrong state manually, or we drop a state without a data migration because we're lazy, our code will not break horribly, just intentionally.

**Comment by rixx** [comment]
Thank you very much!
